File: code/dc/cosmos.pl
% ===================================================================
% cosmos.pl    THE RELATIONAL SPINE
% The DC four-fold network as a NATIVE RELATION.
% Claim R1-R7 ("any two variables determine the other two") is not
% asserted here; it is RUN. Query dc/4 with any two bound, get the rest.
% The AbsentVar is a logic variable. Undecidability is a goal that is
% not proven false -- it is returned as an OPEN INTERFACE.
% ===================================================================

:- use_module(library(clpr)).      % constraints over the reals: the relation runs in every direction

% -- The closed algebra. Four variables, mutually determining. --------
% P = U*I ,  U = I*R  (=> P = I^2 R,  P = U^2/R). No addition. No external constant.
    
dc(P,U,I,R) :-
    { P = U * I },
    { U = I * R }.

% Solve from any pair. We keep it explicit so the "any two determine the
% rest" claim is legible as EXECUTION, not faith.
    
solve(u-U, i-I, P, R) :- P is U*I, ( I =:= 0 -> R = boundary ; R is U/I ).
solve(u-U, r-R, P, I) :- ( R =:= 0 -> I = boundary, P = boundary ; I is U/R, P is U*I ).
solve(i-I, r-R, P, U) :- U is I*R, P is I*I*R.
solve(p-P, r-R, U, I) :- ( R =:= 0 -> U = boundary, I = boundary
                         ; U is sqrt(P*R), I is sqrt(P/R) ).
solve(p-P, i-I, U, R) :- ( I =:= 0 -> U = boundary, R = boundary
                         ; U is P/I, R is P/(I*I) ).
solve(u-U, p-P, I, R) :- ( U =:= 0 -> I = boundary, R = boundary
                         ; I is P/U, R is (U*U)/P ).

% -- The AbsentVar: what is HELD, not solved. -------------------------
% To .behold() the algebra is to leave one variable unbound as live context.
% Prolog does this natively: an unbound variable is a carried absence.
    
beholding(P,U,I,R, Absent) :-
    term_variables([P,U,I,R], Vs),
    ( Vs = [] -> Absent = none            % fully .observe()d: 0-DoF crystal
    ; Absent = held(Vs) ).                % live interior: DoF remain

% -- Undecidability as interface, not as falsehood. -------------------
% The Tangent Singularity: a coordinate that diverges. We do not fail;
% we return the interface the divergence opens.
    
interface(tangent_singularity, "coordinate diverges at 90deg; orthogonal turn available").
interface(absent_var,          "belonging relation held unmanifest; interior volume preserved").
interface(propositional_debt,  "cross-domain mapping asserted, not derived; Landauer tax unpaid").

% A boundary is reached, and REPORTED, rather than swallowed.
report_boundary(Kind) :- interface(Kind, Msg), format("  [INTERFACE ~w] ~w~n", [Kind, Msg]).

% ===================================================================
% DEMONSTRATION
% ===================================================================
demo :-
    format("~n== RELATIONAL SPINE (Prolog) ==~n"),
    format("Same relation, run in four different directions:~n"),

    solve(u-12.0, i-2.0, P1, R1),
    format("  given U=12, I=2   -> P=~w  R=~w   [COMPUTED]~n", [P1, R1]),

    solve(p-100.0, r-4.0, U2, I2),
    format("  given P=100, R=4  -> U=~w  I=~w   [COMPUTED]~n", [U2, I2]),

    solve(i-3.0, r-5.0, P3, U3),
    format("  given I=3, R=5    -> P=~w  U=~w   [COMPUTED]~n", [P3, U3]),

    format("~nThe boundary is not an error. It is an interface:~n"),
    solve(u-12.0, i-0.0, P4, R4),
    format("  given U=12, I=0   -> P=~w  R=~w~n", [P4, R4]),
    report_boundary(tangent_singularity),

    format("~nThe AbsentVar (what the algebra HOLDS rather than solves):~n"),
    beholding(_P, 12.0, 2.0, _R, A1),
    format("  behold(U=12,I=2, P?,R?) -> ~w   [interior has DoF]~n", [A1]),
    beholding(24.0, 12.0, 2.0, 12.0, A2),
    format("  observe(all four bound) -> ~w    [0-DoF crystal]~n", [A2]),
    report_boundary(absent_var),
    format("~nSpine holds. Relation total where total; interfaced where open.~n").

:- initialization((demo, halt)).

    

File: code/dc/cosmos.fs
\ ==================================================================
\ COMPILING REALITY — FORTH RUNTIME (IMPERATIVE SUBSTRATE)
\ ==================================================================
\ Maps the codex ontology to an explicit, verb-first, stack-driven
\ execution model. Forth does not "solve" equations; it performs
\ the collapse. The buffer is memory. The tax is a counter.
\ The boundary is an interface. The singularity is a readout.
\ ==================================================================

\ --- K4 RELATIONAL BUFFER (4-cell memory volume) ---
CREATE K4-BUF 4 CELLS ALLOT
: P@ K4-BUF @ ;          : P! K4-BUF ! ;
: U@ K4-BUF CELL+ @ ;    : U! K4-BUF CELL+ ! ;
: I@ K4-BUF 2 CELLS + @ ;: I! K4-BUF 2 CELLS + ! ;
: R@ K4-BUF 3 CELLS + @ ;: R! K4-BUF 3 CELLS + ! ;

\ --- STATE & BUDGET ---
-1 CONSTANT ABSENT
VARIABLE DOF-COUNT 4 DOF-COUNT !  \ Tracks unbound slots (AbsentVar volume)
VARIABLE LANDAUER-TAX 0 LANDAUER-TAX ! \ Thermodynamic ledger

\ --- INIT: Grow the empty volume (L0-HolographicDescent) ---
: .INIT-REL ( -- )
  ABSENT P! ABSENT U! ABSENT I! ABSENT R!
  4 DOF-COUNT ! 0 LANDAUER-TAX !
  ." [BEHOLD] K4 volume grown. 4-DoF interior established." CR
;

\ --- BIND: Fill a slot, reduce DoF, maintain verb-first discipline ---
\ Codex: AbsentVar is not missing data; it is structural volume.
: >P ( n -- ) DUP ABSENT = IF DROP EXIT THEN DOF-COUNT @ 0> IF P! 1- DOF-COUNT ! THEN ;
: >U ( n -- ) DUP ABSENT = IF DROP EXIT THEN DOF-COUNT @ 0> IF U! 1- DOF-COUNT ! THEN ;
: >I ( n -- ) DUP ABSENT = IF DROP EXIT THEN DOF-COUNT @ 0> IF I! 1- DOF-COUNT ! THEN ;
: >R ( n -- ) DUP ABSENT = IF DROP EXIT THEN DOF-COUNT @ 0> IF R! 1- DOF-COUNT ! THEN ;

\ --- .BEHOLD: Read interior state without collapse (L1-CompilingReality) ---
\ Codex: .behold() maintains the decoupled buffer. It does not force a value.
: .BEHOLD ( -- )
  ." [BEHOLD] State: " DOF-COUNT @ 0 . ." -DoF interior. "
  DOF-COUNT @ CASE
    4 OF ." All vars suspended. Volume open." ENDOF
    3 OF ." One slot bound. Trajectory entering." ENDOF
    2 OF ." Two slots bound. Mutual determination forming." ENDOF
    1 OF ." Three slots bound. Fourth held in reserve." ENDOF
    0 OF ." Volume full. Crystal formed. Ready to .observe()." ENDOF
  ENDCASE CR
;

\ --- LANDAUER TAX: Thermodynamic cost of collapse (L4-AttractorSpace) ---
\ Codex: Observation degrades hardware. Fuel must be consumed.
: .PAY-TAX ( -- )
  1 LANDAUER-TAX +!
  ." [THERMO] Landauer Tax paid. Budget: " LANDAUER-TAX @ . CR
;

\ --- SINGULARITY INTERFACE: Boundary as structural readout (L4-DimensionalCollapse) ---
\ Codex: Tangent divergence is not an error. It is a coordinate pole.
: .INTERFACE-SINGULARITY ( -- )
  ." [INTERFACE] Tangent singularity encountered (I=0 or R=0)." CR
  ." [BOUNDARY] Coordinate system reaches pole. Orthogonal turn available." CR
  ." [NO-CRASH] Divergence reported. Volume preserved." CR
;

\ --- .OBSERVE: Force evaluation, pay tax, apply DC relations (L3-FormalFoundations) ---
\ Codex: .observe() is compelled. Mutual determination P=U*I, U=I*R resolves.
: .OBSERVE ( -- )
  DOF-COUNT @ 0 <> IF
    ." [TRAJECTORY LOSS] .observe() called with " DOF-COUNT @ . ." DoF remaining." CR
    ." [ABORT] System refuses to fake certainty. Hold the buffer." CR
    EXIT
  THEN
  I@ 0= IF .INTERFACE-SINGULARITY EXIT THEN
  R@ 0= IF .INTERFACE-SINGULARITY EXIT THEN
  .PAY-TAX
  \ Apply mutual determination (DC crystal)
  I@ R@ * DUP U!
  I@ OVER * DUP P!
  ." [OBSERVED] Collapse executed. Relations resolved:" CR
  ." P=" P@ . CR ." U=" U@ . CR ." I=" I@ . CR ." R=" R@ . CR
;

\ --- VERIFY: Check consistency without forcing (Proof_AlgebraicMapping) ---
: .VERIFY-REL ( -- flag )
  P@ U@ I@ * = U@ I@ R@ * = AND
;

\ ==================================================================
\ DEMONSTRATION SEQUENCE
\ ==================================================================
: DEMO ( -- )
  CR ." ============================================" CR
  ." INITIALIZING K4 VOLUME" CR
  .INIT-REL
  .BEHOLD

  CR ." BINDING INCREMENTALLY (MAINTAINING VOLUME)" CR
  12 >R  3 >I  4 >U  \ Leave P unbound to hold 1-DoF
  .BEHOLD

  CR ." FORCING COLLAPSE (IMPERATIVE ACTUATION)" CR
  .OBSERVE

  CR ." VERIFICATION (MUTUAL DETERMINATION CHECK)" CR
  .VERIFY-REL IF ." [VERIFIED] DC relations hold." CR ELSE ." [FAILED] Trajectory loss." CR THEN

  CR ." TESTING SINGULARITY INTERFACE" CR
  .INIT-REL
  0 >I  1 >U  1 >R
  .OBSERVE  \ Will trigger interface, not crash

  CR ." THERMODYNAMIC LEDGER STATE" CR
  ." Total Landauer Tax paid: " LANDAUER-TAX @ . CR
  ." ============================================" CR
;
